### PR TITLE
fix: session termination and simplified management

### PR DESCRIPTION
**Pull request by jbutler**
**Issue #, if available:**

**Description of changes:**

**Why is this change necessary:**

**How was this change tested:**

**Any additional information or context required to review the change:**

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: moquette/broker/src/main/java/io/moquette/broker/SessionRegistry.java [hunk 1]
```diff
@@ -145,82 +139,58 @@ private void recreateSessionPool() {
...
+        if (newIsClean || oldSession == null) {
+            final Session newSession = createNewSession(msg, clientId);
+
             postConnectAction = new SessionCreationResult(newSession, CreationModeEnum.CREATED_CLEAN_NEW, false);
 
             // publish the session
-            final Session previous = pool.putIfAbsent(clientId, newSession);
-            final boolean success = previous == null;
+            LOG.trace("case 1, creating new session with CId {}, clean session {}", clientId, newIsClean);
+            final Session previous = pool.put(clientId, newSession);
```
Comment (MikeDombo) [inline moquette/broker/src/main/java/io/moquette/broker/SessionRegistry.java:152]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `get()` and `put()` may not be thread-safe at lines: **143, 145, and 152**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.

 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/startupprogress/StartupProgress.java#L272-L273) (external link).

File: moquette/broker/src/main/java/io/moquette/broker/Session.java [hunk 2]
```diff
@@ -113,6 +113,34 @@ public int compareTo(Delayed o) {
...
+        }
+        disconnect();
+
+        // Release queued messages
+        while (!sessionQueue.isEmpty()) {
+            final SessionRegistry.EnqueuedMessage msg = sessionQueue.remove();
+            msg.release();
+        }
+
+        // Release in flight messages (QoS 1/2)
```
Comment (jbutler) [inline moquette/broker/src/main/java/io/moquette/broker/Session.java:133]: TODO: this can possibly still throw a concurrent modification exception if another message gets delivered while this session is terminating, which could cause connect failures (and the leak of whatever is left in the inflight list)

File: moquette/broker/src/main/java/io/moquette/broker/SessionRegistry.java [hunk 3]
```diff
@@ -145,82 +139,58 @@ private void recreateSessionPool() {
...
-            // case 4
-            LOG.trace("case 4, oldSession with same CId {} still connected, force to close", clientId);
-            oldSession.closeImmediately();
-            //remove(clientId);
-            creationResult = new SessionCreationResult(newSession, CreationModeEnum.DROP_EXISTING, true);
-        }
+            // case 2
+            reactivateSubscriptions(oldSession, username);
 
-        if (creationResult.mode == CreationModeEnum.DROP_EXISTING) {
```
Comment (sushantAtAws) [inline moquette/broker/src/main/java/io/moquette/broker/SessionRegistry.java:207]: This section still seems relevant to me.
If the old session is not disconnected and new session is not clean, we have to drop the existing one. 

Is there some place else you are handling this case?

File: moquette/broker/src/main/java/io/moquette/broker/Session.java [hunk 4]
```diff
@@ -113,6 +113,34 @@ public int compareTo(Delayed o) {
...
+    /**
+     * This method terminates a Session, which involves disconnecting the client,
+     * if one is connected, and releasing all references to unacknowledged publishes.
+     */
+    void terminateSession() {
+        MQTTConnection connection = this.mqttConnection;
+        if (connection != null) {
+            connection.dropConnection();
+        }
+        disconnect();
```
Comment (rbattle) [inline moquette/broker/src/main/java/io/moquette/broker/Session.java:125]: what does `disconnect` do if connection is null?

for that matter, what does `dropConnection` do that wouldn't be done in `disconnect`?

I'm just confused about why you wouldn't do connection management in `disconnect`

File: moquette/broker/src/main/java/io/moquette/broker/Session.java [hunk 5]
```diff
@@ -113,6 +113,34 @@ public int compareTo(Delayed o) {
...
+
+        // Release queued messages
```
Comment (rbattle) [inline moquette/broker/src/main/java/io/moquette/broker/Session.java:127]: can you explain why you want to release queued messages

File: moquette/broker/src/main/java/io/moquette/broker/Session.java (hunk 2 above)
Comment (rbattle) [reply moquette/broker/src/main/java/io/moquette/broker/Session.java:133]: is it `inflightWindow.remove` that throws the modification exception?

File: moquette/broker/src/main/java/io/moquette/broker/SessionRegistry.java [hunk 6]
```diff
@@ -145,82 +139,58 @@ private void recreateSessionPool() {
...
-                pool.put(clientId, newSession);
-            }
+            LOG.trace("case 2, oldSession with same CId {} disconnected", clientId);
+            creationResult = new SessionCreationResult(oldSession, CreationModeEnum.REOPEN_EXISTING, true);
         } else {
-            LOG.debug("Replace session of client with same id");
-            if (!pool.replace(clientId, oldSession, oldSession)) {
-                throw new SessionCorruptedException("old session was already removed");
-            }
+            // case 3
```
Comment (rbattle) [inline moquette/broker/src/main/java/io/moquette/broker/SessionRegistry.java:178]: are these cases supposed to correspond to the old `// case 3`, `// case 4`, etc...?

I guess not since you've moved the `newIsClean` outside of this method

File: moquette/broker/src/main/java/io/moquette/broker/SessionRegistry.java (hunk 3 above)
Comment (rbattle) [reply moquette/broker/src/main/java/io/moquette/broker/SessionRegistry.java:207]: isn't that the case 3 at line 179 in the new code? oldsession would not be null and it is not clean for the new session, so it does 
```
            postConnectAction = reopenExistingSession(clientId, oldSession, username);
```

if the oldSession is not disconnected, it goes to the else branch here

File: moquette/broker/src/main/java/io/moquette/broker/Session.java (hunk 4 above)
Comment (jbutler) [reply moquette/broker/src/main/java/io/moquette/broker/Session.java:125]: `disconnect` is a misnomer. It doesn't actually touch the TCP connection. It just modifies internal Session state. `dropConnection` drops the actual connection.

But now that I look at this again, `dropConnection` can result in (some, but not all scenarios) this method getting called again. Will re-visit this. It's a bit messy

File: moquette/broker/src/main/java/io/moquette/broker/Session.java (hunk 5 above)
Comment (jbutler) [reply moquette/broker/src/main/java/io/moquette/broker/Session.java:127]: Will update

File: moquette/broker/src/main/java/io/moquette/broker/Session.java (hunk 2 above)
Comment (jbutler) [reply moquette/broker/src/main/java/io/moquette/broker/Session.java:133]: Yes. My previous PR that introduces reference counting was meant to address that and prevent concurrent modifications. I'd like to also pull that back in, but I think this is a good improvement as it stands, even if not perfect

File: moquette/broker/src/main/java/io/moquette/broker/SessionRegistry.java (hunk 3 above)
Comment (jbutler) [reply moquette/broker/src/main/java/io/moquette/broker/SessionRegistry.java:207]: Yep, Rob is right. This is only dealing with persistent sessions at this point. So the Session is being resumed, regardless of whether the old client is connected. Old connection is disconnected at line 180

File: moquette/broker/src/main/java/io/moquette/broker/SessionRegistry.java (hunk 6 above)
Comment (jbutler) [reply moquette/broker/src/main/java/io/moquette/broker/SessionRegistry.java:178]: They don't directly translate, I suppose. The old cases aren't clearly defined and IMO it's very difficult to reason about. Whereas the new cases are actually clearly defined.

Old cases:
1. Brand new session. Could be clean or persistent
2. Old session disconnected, new session is clean
3. Old session disconnected, new session is persistent
4. Old session connected, could be clean or persistent

New cases:
1. Session is clean or new
2. Session is persistent, old session is disconnected
3. Session is persistent, old session is connected

Case 1 now encompasses a little bit of old case 1, 2, and 4. New case 2 translates directly to old case 3. New case 3 encompasses half of old case 4.

Comment (jbutler) [reply moquette/broker/src/main/java/io/moquette/broker/SessionRegistry.java:178]: Let me rephrase. The old cases are defined, but it is difficult to correctly handle Session creation. There are only two things that matter when it comes to reconstructing a new session and I believe the new cases better align with this.

1. Is the session clean? If yes, throw away all of the old stuff and create a brand new, empty session. If the old one doesn't exist, you have no choice. Create a new one.
2. Session is persistent and the broker has an old Session. Resume the old session. If there is an old connection, disconnect it.
